cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,388 @@
-	// the static and assigned devices. This ensures that at application
-	// launch the cgroup only has what is currently assigned.
-	sc_must_snprintf(cgroup_file, sizeof(cgroup_file), "%s%s", cgroup_dir,
-			 "devices.deny");
-	write_string_to_file(cgroup_file, "a");
-
-	// add the common devices
-	for (int i = 0; static_devices[i] != NULL; i++)
-		run_snappy_app_dev_add(udev_s, static_devices[i]);
-
-	// add glob for current and future PTY slaves. We unconditionally add
-	// them since we use a devpts newinstance. Unix98 PTY slaves major
-	// are 136-143.
-	// https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt
-	for (unsigned pty_major = 136; pty_major <= 143; pty_major++) {
-		// '/dev/pts/slaves' is only used for debugging and by
-		// /usr/lib/snapd/snap-device-helper to determine if it is a block
-		// device, so just use something to indicate what the
-		// addition is for
-		_run_snappy_app_dev_add_majmin(udev_s, "/dev/pts/slaves",
-					       pty_major, UINT_MAX);
+	if (udev != NULL && *udev != NULL) {
+		udev_unref(*udev);
+		*udev = NULL;
 	}
+}
 
-	// nvidia modules are proprietary and therefore aren't in sysfs and
-	// can't be udev tagged. For now, just add existing nvidia devices to
-	// the cgroup unconditionally (AppArmor will still mediate the access).
-	// We'll want to rethink this if snapd needs to mediate access to other
-	// proprietary devices.
-	//
-	// Device major and minor numbers are described in (though nvidia-uvm
-	// currently isn't listed):
-	// https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt
-	char nv_path[15] = { 0 };	// /dev/nvidiaXXX
-	const char *nvctl_path = "/dev/nvidiactl";
-	const char *nvuvm_path = "/dev/nvidia-uvm";
-	const char *nvidia_modeset_path = "/dev/nvidia-modeset";
+static void sc_cleanup_udev_enumerate(struct udev_enumerate **enumerate)
+{
+	if (enumerate != NULL && *enumerate != NULL) {
+		udev_enumerate_unref(*enumerate);
+		*enumerate = NULL;
+	}
+}
 
-	struct stat sbuf;
+typedef struct sc_cgroup_fds {
+	int devices_allow_fd;
+	int devices_deny_fd;
+	int cgroup_procs_fd;
+} sc_cgroup_fds;
 
-	// /dev/nvidia0 through /dev/nvidia254
-	for (unsigned nv_minor = 0; nv_minor < 255; nv_minor++) {
-		sc_must_snprintf(nv_path, sizeof(nv_path), "/dev/nvidia%u",
-				 nv_minor);
+static sc_cgroup_fds sc_udev_open_cgroup_v1(const char *security_tag)
+{
+	sc_cgroup_fds fds = { -1, -1, -1 };
 
-		// Stop trying to find devices after one is not found. In this
-		// manner, we'll add /dev/nvidia0 and /dev/nvidia1 but stop
-		// trying to find nvidia3 - nvidia254 if nvidia2 is not found.
-		if (stat(nv_path, &sbuf) != 0) {
-			break;
+	/* Open /sys/fs/cgroup */
+	const char *cgroup_path = "/sys/fs/cgroup";
+	int SC_CLEANUP(sc_cleanup_close) cgroup_fd = -1;
+	cgroup_fd = open(cgroup_path,
+			 O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (cgroup_fd < 0 && errno == ENOENT) {
```

Comment:
should this be 

```suggestion
	if (cgroup_fd < 0 || errno == ENOENT) {
```

or even just

```suggestion
	if (cgroup_fd < 0) {
```

? 
I'm a bit confused how we could ever get to the `die("cannot open ...)` code below due to the `&& errno ==ENOENT` and the subsequent `if (errno == ENOENT) {` inside that if.

Reply:
Thanks for spotting this. I'll adjust this code to be less surprising.